pyro/util.py

```diff
@@ -386,6 +387,56 @@ def check_site_shape(site, max_iarange_nesting):
     # TODO Check parallel dimensions on the left of max_iarange_nesting.
 
 
+def _are_independent(counters1, counters2):
+    for name, counter1 in counters1.items():
+        if name in counters2:
+            if counters2[name] != counter1:
+                return True
+    return False
+
+
+def check_traceenum_requirements(model_trace, guide_trace):
+    """
+    Warn if user could easily rewrite the model or guide in a way that would
+    clearly avoid invalid dependencies on enumerated variables.
+
+    :class:`~pyro.infer.traceenum_elbo.TraceEnum_ELBO` enumerates over
+    synchronized products rather than full cartesian products. Therefore models
+    must ensure that no variable outside of an iarange depends on an enumerated
+    variable inside that iarange. Since full dependency checking is impossible,
+    this function aims to warn only in cases where models can be easily
+    rewitten to be obviously correct.
+    """
+    enumerated_sites = set(name for name, site in guide_trace.nodes.items()
+                           if site["type"] == "sample" and site["infer"].get("enumerate"))
+    for role, trace in [('model', model_trace), ('guide', guide_trace)]:
+        irange_counters = {}
+        enumerated_contexts = defaultdict(set)
+        for name, site in trace.nodes.items():
+            if site["type"] != "sample":
+                continue
+            irange_counter = {f.name: f.counter for f in site["cond_indep_stack"] if not f.vectorized}
+            context = frozenset(f for f in site["cond_indep_stack"] if f.vectorized)
```

Comment:
We only care about vectorized frames, since only vectorized variables are enumerated in lock-step.